flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedClientHandler.java

```diff
@@ -0,0 +1,283 @@
+
+			RemoteInputChannel inputChannel = inputChannels.get(bufferOrEvent.receiverId);
+			if (inputChannel == null) {
+				bufferOrEvent.releaseBuffer();
+
+				cancelRequestFor(bufferOrEvent.receiverId);
+
+				return;
+			}
+
+			decodeBufferOrEvent(inputChannel, bufferOrEvent);
+
+		} else if (msgClazz == NettyMessage.ErrorResponse.class) {
+			// ---- Error ---------------------------------------------------------
+			NettyMessage.ErrorResponse error = (NettyMessage.ErrorResponse) msg;
+
+			SocketAddress remoteAddr = ctx.channel().remoteAddress();
+
+			if (error.isFatalError()) {
+				notifyAllChannelsOfErrorAndClose(new RemoteTransportException(
+					"Fatal error at remote task manager '" + remoteAddr + "'.",
+					remoteAddr,
+					error.cause));
+			} else {
+				RemoteInputChannel inputChannel = inputChannels.get(error.receiverId);
+
+				if (inputChannel != null) {
+					if (error.cause.getClass() == PartitionNotFoundException.class) {
+						inputChannel.onFailedPartitionRequest();
+					} else {
+						inputChannel.onError(new RemoteTransportException(
+							"Error at remote task manager '" + remoteAddr + "'.",
+							remoteAddr,
+							error.cause));
+					}
+				}
+			}
+		} else {
+			throw new IllegalStateException("Received unknown message from producer: " + msg.getClass());
+		}
+	}
+
+	private void decodeBufferOrEvent(RemoteInputChannel inputChannel, NettyMessage.BufferResponse bufferOrEvent) throws Throwable {
+		try {
+			if (bufferOrEvent.isBuffer()) {
+				// ---- Buffer ------------------------------------------------
+
+				// Early return for empty buffers. Otherwise Netty's readBytes() throws an
+				// IndexOutOfBoundsException.
+				if (bufferOrEvent.getSize() == 0) {
+					inputChannel.onEmptyBuffer(bufferOrEvent.sequenceNumber);
+					return;
+				}
+
+				Buffer buffer = inputChannel.requestBuffer();
+				if (buffer != null) {
+					buffer.setSize(bufferOrEvent.getSize());
+					bufferOrEvent.getNettyBuffer().readBytes(buffer.getNioBuffer());
+
+					inputChannel.onBuffer(buffer, bufferOrEvent.sequenceNumber);
+					inputChannel.onSenderBacklog(bufferOrEvent.backlog);
+				} else if (inputChannel.isReleased()) {
+					cancelRequestFor(bufferOrEvent.receiverId);
+				} else {
+					throw new IOException("There should always have available buffer for credit-based.");
+				}
+			} else {
+				// ---- Event -------------------------------------------------
+				// TODO We can just keep the serialized data in the Netty buffer and release it later at the reader
+				byte[] byteArray = new byte[bufferOrEvent.getSize()];
+				bufferOrEvent.getNettyBuffer().readBytes(byteArray);
+
+				MemorySegment memSeg = MemorySegmentFactory.wrap(byteArray);
+				Buffer buffer = new Buffer(memSeg, FreeingBufferRecycler.INSTANCE, false);
+
+				inputChannel.onBuffer(buffer, bufferOrEvent.sequenceNumber);
+				inputChannel.onSenderBacklog(bufferOrEvent.backlog);
```

Comment:
It seems that these two calls always belong together. Maybe add `bufferOrEvent.backlog` as the third parameter to `RemoteInputChannel#onBuffer()`?

Reply:
Yes, I can modify for that, but also need to modify it in `PartitionRequestClientHandler`.

Reply:
yes, I guess, you can use an invalid backlog (`-1`?) to differentiate? Don't put too much effort into it since `PartitionRequestClientHandler` will be removed at some point.